### PR TITLE
Bh/lock funcs

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: sql/session.go [hunk 1]
```diff
@@ -177,6 +184,40 @@ func (s *BaseSession) WarningCount() uint16 {
...
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	s.locks[lockName] = true
+	return nil
+}
+
+// DelLock removes a lock from the set of locks owned by this user
+func (s *BaseSession) DelLock(lockName string) error {
+	s.mu.RLock()
```
Comment (zachmu) [inline sql/session.go:198]: Shouldn't this be Lock()?

File: sql/session.go [hunk 2]
```diff
@@ -177,6 +184,40 @@ func (s *BaseSession) WarningCount() uint16 {
 	return uint16(len(s.warnings))
 }
 
+// AddLock adds a lock to the set of locks owned by this user which will need to be released if this session terminates
+func (s *BaseSession) AddLock(lockName string) error {
+	s.mu.RLock()
```
Comment (zachmu) [inline sql/session.go:189]: Shouldn't this be Lock()?

File: sql/lock_subsystem_test.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	err = ls.Lock(user1Ctx, testLockName, 0)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx, testLockName))
+
+	err = ls.Lock(user2Ctx, testLockName, 0)
+	assert.Error(t, err)
+	assert.Nil(t, getLockDiffs(user2Ctx))
+}
+
+func TestRace(t *testing.T) {
```
Comment (zachmu) [inline sql/lock_subsystem_test.go:88]: This is fine, but there's specific go tooling for finding races that you should figure out how to effectively tap into with a test.

File: sql/lock_subsystem.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+			if newVal.Count == 0 {
+				return ctx.Session.DelLock(name)
+			}
+
+			return nil
+		}
+	}
+}
+
+// ReleaseALl releases all locks the ID associated with the given ctx.Session
```
Comment (zachmu) [inline sql/lock_subsystem.go:122]: Add a note about return vals

File: sql/expression/function/locks.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/liquidata-inc/go-mysql-server/sql/expression"
+	"gopkg.in/src-d/go-errors.v1"
+	"strings"
+	"time"
+)
+
+// ErrIllegalLockNameArgType is a kind of error that is thrown when the parameter passed as a lock name is not a string.
+var ErrIllegalLockNameArgType = errors.NewKind("Illegal parameter data type %s for operation '%s'")
+
+// ReleaseAllLocksForLS returns the logic to execute whet the sql function release_all_locks is executed
```
Comment (zachmu) [inline sql/expression/function/locks.go:15]: s/whet/when/

File: sql/expression/function/locks.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	lockName, ok := val.(string)
+
+	if !ok {
+		return nil, ErrIllegalLockNameArgType.New(nl.Child.Type().String(), nl.funcName)
+	}
+
+	return nl.logic(ctx, nl.ls, lockName)
+}
+
+// String implements the Stringer interface.
```
Comment (zachmu) [inline sql/expression/function/locks.go:68]: fmt.Stringer

File: sql/expression/function/locks.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	}
+}
+
+// Eval implements the Expression interface.
+func (gl *GetLock) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	if gl.Left == nil {
+		return nil, nil
+	}
+
+	val0, err := gl.Left.Eval(ctx, row)
```
Comment (zachmu) [inline sql/expression/function/locks.go:152]: leftVal and rightVal preferable to val0, val1

File: sql/expression/function/locks.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	lockName, ok := val0.(string)
+
+	if !ok {
+		return nil, ErrIllegalLockNameArgType.New(gl.Left.Type().String(), "get_lock")
+	}
+
+	timeout, err := sql.Int64.Convert(val1)
+
+	if err != nil {
+		return nil, fmt.Errorf("illegal value for timeeout %v", timeout)
```
Comment (zachmu) [inline sql/expression/function/locks.go:185]: timeout

File: sql/expression/function/locks.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+// IsNullable implements the Expression interface.
+func (gl *GetLock) IsNullable() bool {
+	return false
+}
+
+// WithChildren implements the Expression interface.
+func (gl *GetLock) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 1 {
```
Comment (zachmu) [inline sql/expression/function/locks.go:213]: 2, not 1.

Should be an engine test that fails from this mistake

File: sql/session.go (hunk 1 above)
Comment (bheni) [reply sql/session.go:198]: fixed

File: sql/session.go (hunk 2 above)
Comment (bheni) [reply sql/session.go:189]: fixed

File: sql/lock_subsystem.go (hunk 4 above)
Comment (bheni) [reply sql/lock_subsystem.go:122]: done

File: sql/expression/function/locks.go (hunk 5 above)
Comment (bheni) [reply sql/expression/function/locks.go:15]: fixed

File: sql/expression/function/locks.go (hunk 6 above)
Comment (bheni) [reply sql/expression/function/locks.go:68]: replaced this comment in 12 places

File: sql/expression/function/locks.go (hunk 7 above)
Comment (bheni) [reply sql/expression/function/locks.go:152]: changed

File: sql/expression/function/locks.go (hunk 9 above)
Comment (bheni) [reply sql/expression/function/locks.go:213]: added test coverage for this in function_test and added function tests.

**Pull request merged**